File: src/stores/account.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { accountApi } from '@/lib/tauri'
import type { Account, CreateAccountDto, UpdateAccountDto, AccountStats } from '@/types/common'

export const useAccountStore = defineStore('account', () => {
  const accounts = ref<Account[]>([])
  const currentAccount = ref<Account | null>(null)
  const currentAccountStats = ref<AccountStats | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  const accountCount = computed(() => accounts.value.length)
  const totalBalance = computed(() =>
    accounts.value.reduce((sum, acc) => sum + acc.balance, 0),
  )

  async function fetchAccounts() {
    loading.value = true
    error.value = null
    try {
      accounts.value = await accountApi.getAll()
    } catch (e) {
      error.value = e instanceof Error ? e.message : '获取账户列表失败'
    } finally {
      loading.value = false
    }
  }

  async function fetchAccountById(id: string) {
    loading.value = true
    error.value = null
    try {
      currentAccount.value = await accountApi.getById(id)
    } catch (e) {
      error.value = e instanceof Error ? e.message : '获取账户详情失败'
    } finally {
      loading.value = false
    }
  }

  async function selectAccount(id: string) {
    const account = accounts.value.find((a) => a.id === id)
    if (account) {
      currentAccount.value = account
    } else {
      try {
        currentAccount.value = await accountApi.getById(id)
      } catch (e) {
        error.value = e instanceof Error ? e.message : '获取账户详情失败'
        return
      }
    }
    // Fetch stats for the selected account
    await fetchAccountStats(id)
  }

  async function fetchAccountStats(id: string) {
    try {
      currentAccountStats.value = await accountApi.getStats(id)
    } catch (e) {
      error.value = e instanceof Error ? e.message : '获取账户统计失败'
    }
  }

  async function createAccount(dto: CreateAccountDto) {
    loading.value = true
    error.value = null
    try {
      const newAccount = await accountApi.create(dto)
      accounts.value.push(newAccount)
      return newAccount
    } catch (e) {
      error.value = e instanceof Error ? e.message : '创建账户失败'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updateAccount(dto: UpdateAccountDto) {
    loading.value = true
    error.value = null
    try {
      const updated = await accountApi.update(dto)
      const index = accounts.value.findIndex((a) => a.id === dto.id)
      if (index > -1) {
        accounts.value[index] = updated
      }
      if (currentAccount.value?.id === dto.id) {
        currentAccount.value = updated
      }
      return updated
    } catch (e) {
      error.value = e instanceof Error ? e.message : '更新账户失败'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deleteAccount(id: string) {
    loading.value = true
    error.value = null
    try {
      await accountApi.delete(id)
      accounts.value = accounts.value.filter((a) => a.id !== id)
      if (currentAccount.value?.id === id) {
        currentAccount.value = null
        currentAccountStats.value = null
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : '删除账户失败'
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    accounts,
    currentAccount,
    currentAccountStats,
    loading,
    error,
    accountCount,
    totalBalance,
    fetchAccounts,
    fetchAccountById,
    fetchAccountStats,
    selectAccount,
    createAccount,
    updateAccount,
    deleteAccount,
  }
})

File: src-tauri/src/commands/account.rs
//! 账户相关 Tauri Commands
//!
//! 提供账户管理和交易计算的前端调用接口

use crate::db::DbState;
use crate::error::AppError;
use crate::models::{Account, AccountStats, CreateAccountDto, UpdateAccountDto};
use crate::services::{AccountService, CalculationService};
use crate::services::calculation_service::CalculationResult;
use tauri::State;

/// 获取所有账户列表
#[tauri::command]
pub fn get_accounts(state: State<'_, DbState>) -> Result<Vec<Account>, AppError> {
    AccountService::get_all(&state)
}

/// 根据 ID 获取单个账户
#[tauri::command]
pub fn get_account(state: State<'_, DbState>, id: String) -> Result<Account, AppError> {
    AccountService::get_by_id(&state, &id)
}

/// 创建新账户
#[tauri::command]
pub fn create_account(
    state: State<'_, DbState>,
    dto: CreateAccountDto,
) -> Result<Account, AppError> {
    AccountService::create(&state, &dto)
}

/// 更新账户信息
#[tauri::command]
pub fn update_account(
    state: State<'_, DbState>,
    dto: UpdateAccountDto,
) -> Result<Account, AppError> {
    AccountService::update(&state, &dto)
}

/// 删除账户
#[tauri::command]
pub fn delete_account(state: State<'_, DbState>, id: String) -> Result<(), AppError> {
    AccountService::delete(&state, &id)
}

/// 获取账户统计数据
///
/// 返回指定账户的交易统计信息，包括胜率、盈亏比等
#[tauri::command]
pub fn get_account_stats(
    state: State<'_, DbState>,
    account_id: String,
) -> Result<AccountStats, AppError> {
    AccountService::get_stats(&state, &account_id)
}

/// 执行交易计算
///
/// 根据账户参数和交易价格计算风险指标
#[tauri::command]
pub fn calculate_trade(
    state: State<'_, DbState>,
    account_id: String,
    entry_price: f64,
    stop_loss: f64,
    take_profit: f64,
) -> Result<CalculationResult, AppError> {
    // 获取账户信息
    let account = AccountService::get_by_id(&state, &account_id)?;

    // 执行计算
    let result = CalculationService::calculate(
        account.balance,
        account.risk_ratio,
        account.point_value,
        entry_price,
        stop_loss,
        take_profit,
    );

    Ok(result)
}
